dependency.go

```diff
@@ -0,0 +1,171 @@
+// will be used to pass the DNS name to the parent container.
+func NewContainerDependency(containerRequest ContainerRequest, envKey string) ContainerDependency {
+	return ContainerDependency{
+		Request:      containerRequest,
+		EnvKey:       envKey,
+		CallbackFunc: func(c Container) {},
+		KeepAlive:    true,
+	}
+}
+
+func (c ContainerDependency) WithKeepAlive(keepAlive bool) ContainerDependency {
+	c.KeepAlive = keepAlive
+	return c
+}
+
+func (c ContainerDependency) WithCallback(callbackFunc func(Container)) ContainerDependency {
+	c.CallbackFunc = callbackFunc
+	return c
+}
+
+func (c ContainerDependency) StartDependency(ctx context.Context, network string) (Container, error) {
+	c.Request.Networks = append(c.Request.Networks, network)
+	dependency, err := GenericContainer(ctx, GenericContainerRequest{
+		ContainerRequest: c.Request,
+		Started:          true,
+		Reuse:            c.Request.Name != "", // reuse a running dependency container if a name is provided.
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c.CallbackFunc(dependency)
+	return dependency, nil
+}
+
+func resolveDNSName(ctx context.Context, container Container, network *DockerNetwork) (string, error) {
+	curNetworks, err := container.Networks(ctx)
+	if err != nil {
+		return "", fmt.Errorf("%w: could not retrieve networks for dependency container", err)
+	}
+	// The container may not be connected to the network if it was reused.
+	if slices.Index(curNetworks, network.Name) == -1 {
+		err = network.provider.client.NetworkConnect(ctx, network.ID, container.GetContainerID(), nil)
+		if err != nil {
+			return "", fmt.Errorf("%w: could not connect dependency container to network", err)
+		}
+	}
+
+	networkAlias, err := container.NetworkAliases(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	aliases := networkAlias[network.Name]
+	if len(aliases) == 0 {
+		return "", errors.New("could not retrieve network alias for dependency container")
+	}
+	return aliases[0], nil
+}
+
+func cleanupDependencyNetwork(ctx context.Context, dependencies map[Container]bool, network *DockerNetwork) error {
+	if network == nil {
+		return nil
+	}
+
+	for dependency, keepAlive := range dependencies {
+		err := network.provider.client.NetworkDisconnect(ctx, network.ID, dependency.GetContainerID(), true)
+		if err != nil {
+			return err
+		}
+
+		if !keepAlive {
+			if err := dependency.Terminate(ctx); err != nil {
+				return err
+			}
+		}
+
+	}
+	defer network.provider.Close()
+	return network.Remove(ctx)
+}
+
+var defaultDependencyHook = func(dockerInput *container.Config) ContainerLifecycleHooks {
+	var depNetwork *DockerNetwork
+	depContainers := make(map[Container]bool)
+	return ContainerLifecycleHooks{
+		PreCreates: []ContainerRequestHook{
+			func(ctx context.Context, req ContainerRequest) (err error) {
+				if len(req.DependsOn) == 0 {
+					return nil
+				}
+				defer func() {
+					if err != nil {
```

Comment:
suggestion: invert the condition and return early reducing indentation and making it easier grok.